pkg/gds/models/v1/contacts.go

```diff
@@ -0,0 +1,216 @@
+		i.current = i.contacts[i.index]
+		contact := i.current.contact
+		if contact != nil {
+			if i.email && !ContactHasEmail(contact) {
+				continue
+			}
+			if i.verified {
+				var verified bool
+				var err error
+				if verified, err = ContactIsVerified(contact); err != nil {
+					// If we can't retrieve the contact verification, let the caller
+					// handle the error on Value().
+					i.current.err = err
+					i.index++
+					return true
+				}
+				if !verified {
+					continue
+				}
+			}
+			i.index++
+			return true
+		}
+	}
+	i.current = nil
+	return false
+}
+
+func (i *ContactIterator) Value() (*pb.Contact, string, error) {
+	if i.current != nil {
+		return i.current.contact, i.current.kind, i.current.err
+	}
+	return nil, "", errors.New("no more contacts")
+}
+
+// GetContactVerification token and verified status from the extra data field on the Contact.
+func GetContactVerification(contact *pb.Contact) (_ string, _ bool, err error) {
+	// Return zero-valued defaults with no error if extra is nil.
+	if contact == nil || contact.Extra == nil {
+		return "", false, nil
+	}
+
+	// Unmarshal the extra data field on the Contact
+	extra := &GDSContactExtraData{}
+	if err = contact.Extra.UnmarshalTo(extra); err != nil {
+		return "", false, err
+	}
+	return extra.GetToken(), extra.GetVerified(), nil
+}
+
+// SetContactVerification token and verified status on the Contact record.
+func SetContactVerification(contact *pb.Contact, token string, verified bool) (err error) {
+	if contact == nil {
+		return errors.New("cannot set verification on nil contact")
+	}
+
+	// Unmarshal previous extra data.
+	extra := &GDSContactExtraData{}
+	if contact.Extra != nil {
+		if err = contact.Extra.UnmarshalTo(extra); err != nil {
+			return fmt.Errorf("could not deserialize previous extra: %s", err)
+		}
+	}
+
+	// Set contact verification.
+	extra.Verified = verified
+	extra.Token = token
+	if contact.Extra, err = anypb.New(extra); err != nil {
+		return err
+	}
+	return nil
+}
+
+// VerifiedContacts returns a map of contact type to email address for all verified
+// contacts, omitting any contacts that are not verified or do not exist.
+func VerifiedContacts(vasp *pb.VASP) (contacts map[string]string) {
+	contacts = make(map[string]string)
+	iter := NewContactIterator(vasp.Contacts, false, true)
+	for iter.Next() {
+		if contact, kind, err := iter.Value(); err == nil {
+			contacts[kind] = contact.Email
+		}
+	}
+	return contacts
+}
+
+// ContactVerifications returns a map of contact type to verified status, omitting any
+// contacts that do not exist.
+func ContactVerifications(vasp *pb.VASP) (contacts map[string]bool, err error) {
+	contacts = make(map[string]bool)
+	iter := NewContactIterator(vasp.Contacts, false, false)
+	for iter.Next() {
+		if contact, kind, err := iter.Value(); err == nil {
+			if verified, err := ContactIsVerified(contact); err == nil {
+				contacts[kind] = verified
+			}
+		}
+	}
+	return contacts, nil
+}
```

Comment:
Could we also add a multi-error here, e.g. something similar to the suggested `iter.Error()` above or using https://github.com/hashicorp/go-multierror -- basically in the for loop we have to collect errors, and if an error occurs we have to return the multi-error along with the contact.